<h3>Programming Assignment 2: Data Cleaning</h3>
A research team has been to Antartica and collected some data on penguins. The research team consists of people from different countries, and while they where there during the [midnight sun](https://en.wikipedia.org/wiki/Midnight_sun), the researcher responsible for delivering the data couldn't sleep and made some silly mistakes. You are now hired to fix all the problems and make some simple analysis. 

<h4>Note about the data</h4>
The actual data consists of the [Palmer Penguins dataset](https://github.com/allisonhorst/palmerpenguins/blob/main/README.md) that has been intentionally garbled. All of the mistakes and peculiarities introduced to the data have been something that the course staff has seen in real life datasets.

<h4>Question 1: Read and import libs</h4>
Import pandas and numpy using standard aliases. Read the file `palmer_penguins.csv` and save the result in a variable called `df`. 

The researcher used a Windows computer and stored the data in a non-unicode format, so you will have to address this to correctly read the data.

In [ ]:
import pandas as pd
import numpy as np

df = pd.read_csv('palmer_penguins.csv', encoding='iso-8859-1')
df

Art          Ö  bill.length.mm bill_depth_mm  FlipperLengthMm  \
0       Gentoo     Biscoe            41.7    0,57874047            210.0   
1       Adelie  Torgersen            39.5          39,5            186.0   
2    Chinstrap      Dream            46.9          46,9            192.0   
3       Adelie     Biscoe            36.5    0,65354366            181.0   
4       Gentoo     Biscoe            48.8    0,63779562            222.0   
..         ...        ...             ...           ...              ...   
408  Chinstrap      Dream            55.8          55,8            207.0   
409  Chinstrap      Dream            43.5          43,5            202.0   
410  Chinstrap      Dream            49.6          49,6            193.0   
411  Chinstrap      Dream            50.8          50,8            210.0   
412  Chinstrap      Dream            50.2          50,2            198.0   

     body_mass_g     kön    år  
0         4700.0  female  2009  
1         3800.0  female  2007  
2         2700.0  female  2008  
3         2850.0  female  2008  
4         6000.0    male  2009  
..           ...     ...   ...  
408       4000.0    male  2009  
409       3400.0  female  2009  
410       3775.0    male  2009  
411       4100.0    male  2009  
412       3775.0  female  2009  

[413 rows x 8 columns]

<h4>Question 2: Rename cols to a coherent name style</h4>

Different teams where responsible for collecting data for different variables, and have used different standard. Convert all column names to [snake_case](https://en.wikipedia.org/wiki/Snake_case). Ensure that the column names are in lowercase.

The researcher in charge of the file format was Swedish and for some reason decided to store some columns using Swedish names. Change them into English (species, island, sex, year).

In [ ]:
df.rename(columns={"Art": "species", 
                   "Ö": "island", 
                   "bill.length.mm": "bill_length_mm", 
                   "FlipperLengthMm": "flipper_length_mm", 
                   "kön": "sex", 
                   "år": "year"}, 
          inplace=True)
df

species     island  bill_length_mm bill_depth_mm  flipper_length_mm  \
0       Gentoo     Biscoe            41.7    0,57874047              210.0   
1       Adelie  Torgersen            39.5          39,5              186.0   
2    Chinstrap      Dream            46.9          46,9              192.0   
3       Adelie     Biscoe            36.5    0,65354366              181.0   
4       Gentoo     Biscoe            48.8    0,63779562              222.0   
..         ...        ...             ...           ...                ...   
408  Chinstrap      Dream            55.8          55,8              207.0   
409  Chinstrap      Dream            43.5          43,5              202.0   
410  Chinstrap      Dream            49.6          49,6              193.0   
411  Chinstrap      Dream            50.8          50,8              210.0   
412  Chinstrap      Dream            50.2          50,2              198.0   

     body_mass_g     sex  year  
0         4700.0  female  2009  
1         3800.0  female  2007  
2         2700.0  female  2008  
3         2850.0  female  2008  
4         6000.0    male  2009  
..           ...     ...   ...  
408       4000.0    male  2009  
409       3400.0  female  2009  
410       3775.0    male  2009  
411       4100.0    male  2009  
412       3775.0  female  2009  

[413 rows x 8 columns]

<h4>Question 3: Missing values</h4>
It seems like some of the researchers used SPSS and a standard where unknown values are coded as `-999.0`. 

Replace all the `-999.0` with the numpy variant of non-values, `np.nan`. 

In [ ]:
df.replace(-999.0, np.nan, inplace=True)

df

species     island  bill_length_mm bill_depth_mm  flipper_length_mm  \
0       Gentoo     Biscoe            41.7    0,57874047              210.0   
1       Adelie  Torgersen            39.5          39,5              186.0   
2    Chinstrap      Dream            46.9          46,9              192.0   
3       Adelie     Biscoe            36.5    0,65354366              181.0   
4       Gentoo     Biscoe            48.8    0,63779562              222.0   
..         ...        ...             ...           ...                ...   
408  Chinstrap      Dream            55.8          55,8              207.0   
409  Chinstrap      Dream            43.5          43,5              202.0   
410  Chinstrap      Dream            49.6          49,6              193.0   
411  Chinstrap      Dream            50.8          50,8              210.0   
412  Chinstrap      Dream            50.2          50,2              198.0   

     body_mass_g     sex  year  
0         4700.0  female  2009  
1         3800.0  female  2007  
2         2700.0  female  2008  
3         2850.0  female  2008  
4         6000.0    male  2009  
..           ...     ...   ...  
408       4000.0    male  2009  
409       3400.0  female  2009  
410       3775.0    male  2009  
411       4100.0    male  2009  
412       3775.0  female  2009  

[413 rows x 8 columns]

<h4>Question 4: Type conversion and inch to mm</h4>

It seems like the americans on Biscoe island recorded their measurements on bill depth in inches, while everyone else took proper measurements in mm. Somehow, it seems like this mistake also converted all the measurements to text strings with commas instead of dots. 

First of all, change all commas to dots in `bill_depth_mm`, then change the type to a numeric.

Then write a function `clean_bill_depth` that takes a single row of the dataframe and, if the value is from the island of `Biscoe`, it converts `bill_depth_mm` from inches to mm and returns it. Otherwise it returns `bill_depth_mm` as is. Then `apply` this function to convert all data into millimeters. 

1 inch is by definition exactly 25.4 mm.

In [ ]:
df['bill_depth_mm'] = df['bill_depth_mm'].replace(to_replace=',', value='.', regex=True)

df['bill_depth_mm'] = pd.to_numeric(df['bill_depth_mm'], errors='coerce')

def clean_bill_depth(row):
    # implement this function
    if row['island'] == 'Biscoe':
        return row['bill_depth_mm'] * 25.4
    else:
        return row['bill_depth_mm']

df['bill_depth_mm'] = df.apply(clean_bill_depth, axis=1)

df

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Gentoo     Biscoe            41.7      14.700008              210.0   
1       Adelie  Torgersen            39.5      39.500000              186.0   
2    Chinstrap      Dream            46.9      46.900000              192.0   
3       Adelie     Biscoe            36.5      16.600009              181.0   
4       Gentoo     Biscoe            48.8      16.200009              222.0   
..         ...        ...             ...            ...                ...   
408  Chinstrap      Dream            55.8      55.800000              207.0   
409  Chinstrap      Dream            43.5      43.500000              202.0   
410  Chinstrap      Dream            49.6      49.600000              193.0   
411  Chinstrap      Dream            50.8      50.800000              210.0   
412  Chinstrap      Dream            50.2      50.200000              198.0   

     body_mass_g     sex  year  
0         4700.0  female  2009  
1         3800.0  female  2007  
2         2700.0  female  2008  
3         2850.0  female  2008  
4         6000.0    male  2009  
..           ...     ...   ...  
408       4000.0    male  2009  
409       3400.0  female  2009  
410       3775.0    male  2009  
411       4100.0    male  2009  
412       3775.0  female  2009  

[413 rows x 8 columns]

<h4>Question 5: Remove duplicates</h4>
The researcher who sent you the data seem to have been awake 48 hours straight, and was afraid that some data would be lost when they emailed it to you. Therefore, they duplicated some random rows. 

Find and delete all duplicate rows. 

In [ ]:
df = df.drop_duplicates()
df

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Gentoo     Biscoe            41.7      14.700008              210.0   
1       Adelie  Torgersen            39.5      39.500000              186.0   
2    Chinstrap      Dream            46.9      46.900000              192.0   
3       Adelie     Biscoe            36.5      16.600009              181.0   
4       Gentoo     Biscoe            48.8      16.200009              222.0   
..         ...        ...             ...            ...                ...   
408  Chinstrap      Dream            55.8      55.800000              207.0   
409  Chinstrap      Dream            43.5      43.500000              202.0   
410  Chinstrap      Dream            49.6      49.600000              193.0   
411  Chinstrap      Dream            50.8      50.800000              210.0   
412  Chinstrap      Dream            50.2      50.200000              198.0   

     body_mass_g     sex  year  
0         4700.0  female  2009  
1         3800.0  female  2007  
2         2700.0  female  2008  
3         2850.0  female  2008  
4         6000.0    male  2009  
..           ...     ...   ...  
408       4000.0    male  2009  
409       3400.0  female  2009  
410       3775.0    male  2009  
411       4100.0    male  2009  
412       3775.0  female  2009  

[344 rows x 8 columns]

In this case, a random subset of the dataset was accidentally appended to the end, and they were true duplications of data. Should duplicates always be dropped? Is there a reason to delete some of the data you collected? Give a few examples of when deleting data is ok, and when it might be wrong to do so. 

There are a few situations where it is okay to delete the data. In situations where the data set has true duplicates, a certain human or technical error has caused the data to be replicated - such as in this case - the duplicated data can be safely removed. Likewise, if the data has anomalies that might skew the results or analysis, removing such data is the best practice. However, if for some reason it cannot be determined whether the data points collected are unique (such as an anonymous survey entry), then the data collected should not be removed.

<h4>Question 6: Store the cleaned data</h4>
Create a file called `palmer_penguins_cleaned.csv` that contains the cleaned dataset. Use UTF-8 encoding for the data, full stops (.) for decimal separators, commas as field separators, and use UNIX line breaks (`\n`). Include the correct, lowercase snake_case field names. Do not include the index (the line number). When storing floating-point values, only use 1 decimal digit of precision (e.g., instead of 14.70008, store 14.7).

In [ ]:
df.columns = ['species', 'island', 'bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g', 'sex', 'year']

df.to_csv('palmer_penguins_cleaned.csv', sep=',', index=False, encoding='utf-8', lineterminator='\n', decimal='.', float_format='%.1f')

<h4>Question 7: Custom summary function</h4>
Now the data is *finally* ready for some analysis.

Write a function that named `five_summary` that takes a list of numbers (we consider a Python list, a NumPy ndarray, or a Pandas series a list in this sense), and produces a dictionary as an output, containing the following fields:
- `minimum` containing the smallest value in the list,
- `maximum` containing the largest value in the list,
- `median` containing the median value of the list,
- `mean` containing the mean value of the list, and
- `sd` containing the sample standard deviation of the list. 

`Apply` this function to all all numeric columns of the dataset. Save the result in the variable `summary`. The result should be a `Series` that has the column names as the index and each element is a dictionary of the given form.

### Hints:
- The method `select_dtypes` of the dataframe can be used to select all columns of a specific datatype. `np.number` is a datatype that represents all numeric datatypes.
- Numpy comes with appropriate functions to compute these values, look them up!
- As there are missing values (NaNs), you need to handle these correctly (that is, ignore the values); NumPy provides variants of the basic functions for computing, e.g., mean, such that it ignores NaNs, look it up!
- For computing the standard deviation, use the argument `ddof=1` to specify 1 degree of freedom (this is because we use the sample mean to estimate the average; we will discuss why this correction is done later in class).
- The function `five_summary` should be able to handle also other data, as long as it is numerical; the automatic grader will test this on other data as well.

In [ ]:
def five_summary(column):
    column = np.array(column)
    
    summary = {
        'minimum': np.nanmin(column),  
        'maximum': np.nanmax(column),  
        'median': np.nanmedian(column),  
        'mean': np.nanmean(column),  
        'sd': np.nanstd(column, ddof=1)  
    }
    return summary

numeric_df = df.select_dtypes(include=[np.number])
    
summary = numeric_df.apply(five_summary, axis=0)
    
summary

bill_length_mm       {'minimum': 32.1, 'maximum': 59.6, 'median': 4...
bill_depth_mm        {'minimum': 13.100007073999999, 'maximum': 58....
flipper_length_mm    {'minimum': 172.0, 'maximum': 231.0, 'median':...
body_mass_g          {'minimum': 2700.0, 'maximum': 6300.0, 'median...
year                 {'minimum': 2007, 'maximum': 2009, 'median': 2...
dtype: object

<h4>Question 8: Grouped aggregate by the island</h4>

The result above is somewhat difficult to read and follow. Furthermore, since we are aggregating values from different populations (different species of penguins living on different island), it might not be the best course of action.

We are going to repeat this idea, but construct a hierarchical index where we group the penguins by the island they live in, then by the species, and then both. We are going to use the `agg` method to construct the same summary as before (`min`, `max`, `median`,`mean`,`std`). Group on island, species, and both. 

Save the answer in the variable `result_island`. The resulting dataframe should have a two-level hierarchical index where the first level corresponds to the column `island` and the second is labelled `variable` and contains the numerical column names. There should be 5 columns called `min`, `max`, `median`, `mean`, and `std`, and each row should contain these summary statistics for that particular column for those rows whose `island` corresponds to the value to the left.

That is, for example, if `island` is `Biscoe` and variable is `bill_depth_mm`, the value in column `mean` should be the mean of all `bill_depth_mm` values of those rows where `island` is `Biscoe`.

#### Hints:
- `pd.melt` is very useful, as we can then simply identify measurements by the species, island, sex, and variable in question, and this works nicely with `agg`.
- `drop` can be used to remove columns that are in the way for some operations.
- `groupby` is useful for creating the desired hierarchical index.
- `agg` can take a list of strings of the most commonly used operations.


In [ ]:
melted_df = pd.melt(df, id_vars=['island'], 
                    value_vars=['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g', 'year'],
                    var_name='variable', value_name='value')

grouped = melted_df.groupby(['island', 'variable']).agg(
    min=('value', 'min'),
    max=('value', 'max'),
    median=('value', 'median'),
    mean=('value', 'mean'),
    std=('value', 'std')
)

result_island = pd.concat({'value': grouped}, axis=1)
result_island 

value                            \
                                     min          max       median   
island    variable                                                   
Biscoe    bill_depth_mm        13.100007    21.100011    15.500008   
          bill_length_mm       34.500000    59.600000    45.800000   
          body_mass_g        2850.000000  6300.000000  4775.000000   
          flipper_length_mm   172.000000   231.000000   214.000000   
          year               2007.000000  2009.000000  2008.000000   
Dream     bill_depth_mm        32.100000    58.000000    44.650000   
          bill_length_mm       32.100000    58.000000    44.650000   
          body_mass_g        2700.000000  4800.000000  3687.500000   
          flipper_length_mm   178.000000   212.000000   193.000000   
          year               2007.000000  2009.000000  2008.000000   
Torgersen bill_depth_mm        33.500000    46.000000    38.900000   
          bill_length_mm       33.500000    46.000000    38.900000   
          body_mass_g        2900.000000  4700.000000  3700.000000   
          flipper_length_mm   176.000000   210.000000   191.000000   
          year               2007.000000  2009.000000  2008.000000   

                                                      
                                    mean         std  
island    variable                                    
Biscoe    bill_depth_mm        15.874859    1.820722  
          bill_length_mm       45.257485    4.772731  
          body_mass_g        4716.017964  782.855743  
          flipper_length_mm   209.706587   14.142268  
          year               2008.095238    0.783345  
Dream     bill_depth_mm        44.167742    5.953527  
          bill_length_mm       44.167742    5.953527  
          body_mass_g        3712.903226  416.644112  
          flipper_length_mm   193.072581    7.507366  
          year               2007.983871    0.855246  
Torgersen bill_depth_mm        38.950980    3.025318  
          bill_length_mm       38.950980    3.025318  
          body_mass_g        3706.372549  445.107940  
          flipper_length_mm   191.196078    6.232238  
          year               2007.923077    0.836570

<h4>Question 9: Grouped aggregate by the species</h4>
Do the same but group by the species. Save the answer in the variable `result_species`.

In [ ]:
melted_df = pd.melt(df, id_vars=['species'], 
                    value_vars=['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g', 'year'],
                    var_name='variable', value_name='value')

grouped = melted_df.groupby(['species', 'variable']).agg(
    min=('value', 'min'),
    max=('value', 'max'),
    median=('value', 'median'),
    mean=('value', 'mean'),
    std=('value', 'std')
)

result_species = pd.concat({'value': grouped}, axis=1)
result_species

value                            \
                                     min          max       median   
species   variable                                                   
Adelie    bill_depth_mm        16.000009    46.000000    37.000000   
          bill_length_mm       32.100000    46.000000    38.800000   
          body_mass_g        2850.000000  4775.000000  3700.000000   
          flipper_length_mm   172.000000   210.000000   190.000000   
          year               2007.000000  2009.000000  2008.000000   
Chinstrap bill_depth_mm        40.900000    58.000000    49.550000   
          bill_length_mm       40.900000    58.000000    49.550000   
          body_mass_g        2700.000000  4800.000000  3700.000000   
          flipper_length_mm   178.000000   212.000000   196.000000   
          year               2007.000000  2009.000000  2008.000000   
Gentoo    bill_depth_mm        13.100007    17.300009    15.000008   
          bill_length_mm       40.900000    59.600000    47.300000   
          body_mass_g        3950.000000  6300.000000  5000.000000   
          flipper_length_mm   203.000000   231.000000   216.000000   
          year               2007.000000  2009.000000  2008.000000   

                                                      
                                    mean         std  
species   variable                                    
Adelie    bill_depth_mm        32.787420    9.579184  
          bill_length_mm       38.791391    2.663405  
          body_mass_g        3700.662252  458.566126  
          flipper_length_mm   189.953642    6.539457  
          year               2008.013158    0.821780  
Chinstrap bill_depth_mm        48.833824    3.339256  
          bill_length_mm       48.833824    3.339256  
          body_mass_g        3733.088235  384.335081  
          flipper_length_mm   195.823529    7.131894  
          year               2007.970588    0.863360  
Gentoo    bill_depth_mm        14.982122    0.981220  
          bill_length_mm       47.504878    3.081857  
          body_mass_g        5076.016260  504.116237  
          flipper_length_mm   217.186992    6.484976  
          year               2008.080645    0.792206

<h4>Question 10: Grouped aggregate by the species and the island</h4>
Do the same but group by the species and island. The first level of the index should be species, the second island, and the third variable. Save the answer in the variable `result_both`.


In [ ]:
melted_df = pd.melt(df, id_vars=['species', 'island'], 
                    value_vars=['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g', 'year'],
                    var_name='variable', value_name='value')

grouped = melted_df.groupby(['species','island','variable']).agg(
    min=('value', 'min'),
    max=('value', 'max'),
    median=('value', 'median'),
    mean=('value', 'mean'),
    std=('value', 'std')
)

result_both = pd.concat({'value': grouped}, axis=1)
result_both

value                            \
                                               min          max       median   
species   island    variable                                                   
Adelie    Biscoe    bill_depth_mm        16.000009    21.100011    18.450010   
                    bill_length_mm       34.500000    45.600000    38.700000   
                    body_mass_g        2850.000000  4775.000000  3750.000000   
                    flipper_length_mm   172.000000   203.000000   189.500000   
                    year               2007.000000  2009.000000  2008.000000   
          Dream     bill_depth_mm        32.100000    44.100000    38.550000   
                    bill_length_mm       32.100000    44.100000    38.550000   
                    body_mass_g        2900.000000  4650.000000  3575.000000   
                    flipper_length_mm   178.000000   208.000000   190.000000   
                    year               2007.000000  2009.000000  2008.000000   
          Torgersen bill_depth_mm        33.500000    46.000000    38.900000   
                    bill_length_mm       33.500000    46.000000    38.900000   
                    body_mass_g        2900.000000  4700.000000  3700.000000   
                    flipper_length_mm   176.000000   210.000000   191.000000   
                    year               2007.000000  2009.000000  2008.000000   
Chinstrap Dream     bill_depth_mm        40.900000    58.000000    49.550000   
                    bill_length_mm       40.900000    58.000000    49.550000   
                    body_mass_g        2700.000000  4800.000000  3700.000000   
                    flipper_length_mm   178.000000   212.000000   196.000000   
                    year               2007.000000  2009.000000  2008.000000   
Gentoo    Biscoe    bill_depth_mm        13.100007    17.300009    15.000008   
                    bill_length_mm       40.900000    59.600000    47.300000   
                    body_mass_g        3950.000000  6300.000000  5000.000000   
                    flipper_length_mm   203.000000   231.000000   216.000000   
                    year               2007.000000  2009.000000  2008.000000   

                                                                
                                              mean         std  
species   island    variable                                    
Adelie    Biscoe    bill_depth_mm        18.370464    1.188821  
                    bill_length_mm       38.975000    2.480916  
                    body_mass_g        3709.659091  487.733722  
                    flipper_length_mm   188.795455    6.729247  
                    year               2008.136364    0.765261  
          Dream     bill_depth_mm        38.501786    2.465359  
                    bill_length_mm       38.501786    2.465359  
                    body_mass_g        3688.392857  455.146437  
                    flipper_length_mm   189.732143    6.585083  
                    year               2008.000000    0.852803  
          Torgersen bill_depth_mm        38.950980    3.025318  
                    bill_length_mm       38.950980    3.025318  
                    body_mass_g        3706.372549  445.107940  
                    flipper_length_mm   191.196078    6.232238  
                    year               2007.923077    0.836570  
Chinstrap Dream     bill_depth_mm        48.833824    3.339256  
                    bill_length_mm       48.833824    3.339256  
                    body_mass_g        3733.088235  384.335081  
                    flipper_length_mm   195.823529    7.131894  
                    year               2007.970588    0.863360  
Gentoo    Biscoe    bill_depth_mm        14.982122    0.981220  
                    bill_length_mm       47.504878    3.081857  
                    body_mass_g        5076.016260  504.116237  
                    flipper_length_mm   217.186992    6.484976  
                    year               2008.08064

<h4>Question 11: Discussion</h4>
Now that you can look at some easy accessible summaries of the different typ of penguins on different islands, what can you notice about them? Summarise some of your observations about similarities and differences in one or two paragraphs. 

The data looks at three different species of penguins, located across three different islands in the Antarctic. It seems that the Adelie species is quite common and can be found across in all three islands. On average, they weigh about 3700g, with a flipper length around 190mm. What is fascinating about this species is that, while they have around the same average bill length, body mass and flipper length, the average bill depth for Adelie penguins living in Biscoe island is almost half the size that of the ones living in the other two islands. This could be due to a multitude of reasons, the penguins on that island could have simply just evolved. 

Chinstrap species can only be found in the island of Dream, they weigh an average of 3733g with a flipper length of 196mm. 

Gentoo species can only be found in the island of Biscoe, they weigh an average of 5079g with a flipper length of 217mm. It seems that Gentoo is the largest penguin species in the data, where both their body mass and flipper are greater. However, an interesting observation noticed was that, while they are bigger in size and have bigger flipper, their bill depth is quite small, with an average of 14.9. 

It seems that both the species living in Biscoe island (Adelie & Gentoo) regardless of their size, seem to have very small bill depth. This could be due specific hunting situation on the island, perhaps it is easier to hunt if they have a smaller bill depth. 

Looking at the data as a whole, it can be said that the body mass values of all the penguins are quite far from the mean as they have a consistently high std value. 